clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetch.java

```
@@ -123,6 +125,29 @@ public void acknowledge(final ConsumerRecord<K, V> record, AcknowledgeType type)
         throw new IllegalStateException("The record cannot be acknowledged.");
     }
 
+    /**
+     * Acknowledge a single record by its topic, partition and offset in the current batch.
+     *
+     * @param topic     The topic of the record to acknowledge
+     * @param partition The partition of the record
+     * @param offset    The offset of the record
+     * @param type      The acknowledgment type which indicates whether it was processed successfully
+     */
+    public void acknowledge(final String topic, final int partition, final long offset, final AcknowledgeType type) {
+        for (Map.Entry<TopicIdPartition, ShareInFlightBatch<K, V>> tipBatch : batches.entrySet()) {
+            TopicIdPartition tip = tipBatch.getKey();
+            KafkaException shareException = tipBatch.getValue().getException();
+            if (tip.topic().equals(topic) && (tip.partition() == partition) &&
+                shareException instanceof RecordDeserializationException &&
```

I don't like the `instanceof` and then the cast. One way around this I suppose would be to  `ShareInFlightBatch` have an exception which has a known type, `ShareFetchException` I guess. Then `ShareFetchException` could have methods which extract the topic, partition and offset from the `RecordDeserialization` exception, and then you just need `shareException.offset() == offset)`. wdyt?

---

`ShareFetchException` was actually created to address the issue mentioned in the comments of [KAFKA-19471](https://issues.apache.org/jira/browse/KAFKA-19471,) preventing exceptions from being swallowed. Therefore, the cause could be either a `SerializationException` or a `CorruptRecordException`. Consequently, if we use `ShareFetchException`, we might still need to perform `instanceof` checks followed by casting.

---

Yes, I was thinking about this case last night and you're absolutely correct. But still, `ShareFetchException` could perhaps be able to return an offset if appropriate rather than using `instanceof` and casting. There's got to be a more elegant, OO way of achieving this.